python/pyspark/tests.py

```diff
@@ -2381,6 +2382,34 @@ def test_startTime(self):
         with SparkContext() as sc:
             self.assertGreater(sc.startTime, 0)
 
+    def test_forbid_insecure_gateway(self):
+        # By default, we fail immediately if you try to create a SparkContext
+        # with an insecure gateway
+        gateway = _launch_gateway(insecure=True)
+        with self.assertRaises(Exception) as context:
+            SparkContext(gateway=gateway)
+        self.assertIn("insecure py4j gateway", context.exception.message)
+        self.assertIn("spark.python.allowInsecurePy4j", context.exception.message)
+        self.assertIn("removed in Spark 3.0", context.exception.message)
+
+    def test_allow_insecure_gateway_with_conf(self):
+        with SparkContext._lock:
+            SparkContext._gateway = None
+            SparkContext._jvm = None
```

Comment:
this part of the test really bothers me, so I'd like to explain to reviewers.  Without this, the test passes -- but it passes even without the changes to the main code!  Or rather, it only passes when its run as part of the entire suite, it would fail when run individually.

What's happening is that `SparkContext._gateway` and `SparkContext._jvm` don't get reset by most tests (eg., they are not reset in `sc.stop()`), so a test running before this one will set those variables, and then this test will end up holding on to a gateway which *does* have the `auth_token` set, and so the accumulator server would still work.

Now that in itself sounds crazy to me, and seems like a problem for things like Zeppelin.  I tried just adding these two lines into `sc.stop()`, but then when I ran all the tests, I got a lot of ` java.io.IOException: error=23, Too many open files in system`.  So maybe something else is not getting properly cleaned up properly in the pyspark tests?

I was hoping somebody else might have some ideas about what is going on or if there is a better way to do this.

Reply:
I usually have made separate classes in this case it's closely related with stop and start tho .. (for instance, https://github.com/apache/spark/blob/master/python/pyspark/sql/tests/test_session.py#L27-L73 and https://github.com/apache/spark/blob/master/python/pyspark/sql/tests/test_dataframe.py#L680-L682

Reply:
I don't think that's really answering my question.  I don't have a problem calling start & stop, I'm wondering why `SparkContext._gateway` and `SparkContext._jvm` don't get reset in `sc.stop()`.  This means that if you have multiple spark contexts in one python session (as we do in our tests), they all reuse the same gateway:

https://github.com/apache/spark/blob/4d693ac904d89b3afeba107eb0480120daf78174/python/pyspark/context.py#L300-L302

for normal use of spark, that's not a problem; but it seems like it would be a problem (a) in our tests and (b) for systems like zeppelin, that might have multiple spark contexts over the lifetime of the python session (I assume, anyway ...)

Reply:
Ah, gotya. In that case, we could consider simply move the test class to the top in a separate class as well but .. yes I suspect tests depending on its order isn't a great idea in a way as well. I'm okay as long as the tests pass. I can take a separate look for this later.

Reply:
Yeah this logic is sort rough, in spark-testing-base for example in between tests where folks do not intend to reuse the same Spark context we also clear some extra properties (although we do reuse the gateway). I think for environments where folks want multiple SparkContexts from Python on the same machine they end up using multiple Python processes anyways.

Reply:
ok its good to get some confirmation of this weird behavior ... but I feel like I still don't understand why we don't reset `SparkContext._gateway` and `SparkContext._jvm` in `sc.stop()`; and why when I tried to make that change, I hit all those errors.  if nothing else, any chance this is related to general flakiness in pyspark tests?